### PR TITLE
General first stage

### PR DESCRIPTION
**Pull request by Shloyem**
Just a general review to see if this is the right direction.
Also if you know of a different better mechanism to use out of the box, would like to know

### CODE REVIEW

File: contracts/staking/fuse/FuseStakingV4.sol [hunk 1]
```diff
@@ -152,7 +153,20 @@ contract FuseStakingV4 is Initializable, OwnableUpgradeable, DSMath {
...
 	}
 
-	function _stake(address _to, address _validator, uint256 _amount) internal returns (bool) {
+	function stake(uint _giveBackRatio) public payable returns (bool) {
+		return stake(address(0), DEFAULT_GIVEBACK_RATIO);
+	}
+
+	function stake(address _validator) public payable returns (bool) {
+		return stake(_validator, DEFAULT_GIVEBACK_RATIO);
+	}
```
Comment (sirpy) [inline contracts/staking/fuse/FuseStakingV4.sol:162]: givebackratio should be explicit

File: contracts/staking/fuse/FuseStakingV4.sol [hunk 2]
```diff
@@ -170,14 +184,12 @@ contract FuseStakingV4 is Initializable, OwnableUpgradeable, DSMath {
 			"validator not in approved list"
 		);
 
+		require(_giveBackRatio >= 10000, "giveback should be higher or equal to 10 percent");
```
Comment (sirpy) [inline contracts/staking/fuse/FuseStakingV4.sol:187]: 10000 should be a parameter

File: contracts/staking/fuse/FuseStakingV4.sol [hunk 3]
```diff
@@ -170,16 +189,45 @@ contract FuseStakingV4 is Initializable, OwnableUpgradeable, DSMath {
...
 	}
 
+	/**
+     * @dev Calculates the weighted average of two values based on their weights.
+     * @param valueA The amount for value A
+     * @param weightA The weight to use for value A
+     * @param valueB The amount for value B
+     * @param weightB The weight to use for value B
+     */
+    function weightedAverage(
```
Comment (Shloyem) [inline contracts/staking/fuse/FuseStakingV4.sol:222]: Do we have our own math or UTIL lib to add it there? I see there's DSMath but it doesn't use SafeMathUpgradeable

File: contracts/staking/fuse/FuseStakingV4.sol [hunk 4]
```diff
@@ -217,7 +265,22 @@ contract FuseStakingV4 is Initializable, OwnableUpgradeable, DSMath {
...
+	 * @param to the receiver
+	 * @param from the sender
+	 * @return receiver average giveback ratio if he has one, otherwise sender giveback ratio
+	 */
+	function getTransferGivebackRatio(address to, address from) internal view returns (uint256){
+		return stakersGivebacks[to].avgRatio > 0 ?
+					stakersGivebacks[to].avgRatio :
+					stakersGivebacks[from].avgRatio > 0 ?
+						stakersGivebacks[from].avgRatio :
+						DEFAULT_GIVEBACK_RATIO;
```
Comment (Shloyem) [inline contracts/staking/fuse/FuseStakingV4.sol:283]: Generally there shouldn't be a situation where the sender doesn't have a giveback ratio and we shouldn't get to this line. Can keep as is or throw

**Pull request merged**